spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/DefaultSource.scala

```diff
@@ -0,0 +1,93 @@
+package org.apache.spark.sql.cassandra
+
+import org.apache.spark.sql.{DataFrame, SaveMode, SQLContext}
+import org.apache.spark.sql.SaveMode._
+import org.apache.spark.sql.sources.{CreatableRelationProvider, SchemaRelationProvider, BaseRelation, RelationProvider}
+import org.apache.spark.sql.types.{DataType, StructType}
+
+
+class DefaultSource extends RelationProvider with SchemaRelationProvider with CreatableRelationProvider {
+
+  /** Creates a new relation for a cassandra table given table, keyspace, and cluster as a parameter.*/
+  override def createRelation(sqlContext: SQLContext,
+                              parameters: Map[String, String]): BaseRelation = {
+    val schemaJsonString = getUserDefinedSchema(parameters)
+    val schema = if (schemaJsonString == None) None
+    else Option(DataType.fromJson(schemaJsonString.get).asInstanceOf[StructType])
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), schema)
+  }
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster and schema as a parameter.*/
+  override def createRelation(sqlContext: SQLContext,
+                              parameters: Map[String, String],
+                              schema: StructType): BaseRelation = {
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), Option(schema))
+  }
+
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster, schema and data as a parameter.*/
+  override def createRelation(sqlContext: SQLContext,
+                              mode: SaveMode,
+                              parameters: Map[String, String],
+                              data: DataFrame): BaseRelation = {
+    val table = sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), None)
+
+    mode match {
+      case Append => table.insert(data, false)
+      case Overwrite => throw new UnsupportedOperationException("Overwriting a Cassandra Table is not allowed.")
+      case ErrorIfExists =>
+        if (!table.buildScan().isEmpty()) {
+          throw new UnsupportedOperationException("Writing to a none-empty Cassandra Table is not allowed.")
+        } else {
+          table.insert(data, false)
+        }
+      case Ignore =>
+    }
+
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), None)
+  }
+
+
+  import DefaultSource._
+
+  private def getTable(parameters: Map[String, String]) : String =
```

Comment:
Yup, however this time is not a java-bean pattern. Just a method for getting something from the config. Actually Spark is doing the same in many places. In this case, `getTable` is somehow more readable than simply `table`. And BTW, this should be `tableName`, not `table`... 
Anyway, I'd write this code slightly differently - I'd introduce a wrapper over parameters map and provided fields like "tableName", "keyspaceName", etc.